x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.security.authz.privilege;
+
+import org.apache.lucene.util.automaton.Automaton;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.xpack.core.security.support.Automatons;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.xpack.core.security.support.Automatons.patterns;
+
+/**
+ * This class helps resolving set of cluster privilege names to a {@link Tuple} of {@link ClusterPrivilege} or
+ * {@link ConditionalClusterPrivilege}
+ */
+public final class ClusterPrivilegeResolver {
+
+    public static final Predicate<String> ACTION_MATCHER = DefaultClusterPrivilege.ALL.predicate();
+
+    private static final ConcurrentHashMap<Set<String>, Tuple<ClusterPrivilege, Set<ConditionalClusterPrivilege>>> CACHE =
+            new ConcurrentHashMap<>();
+
+    /**
+     * For given set of privilege names returns a tuple of {@link ClusterPrivilege} and set of predefined fixed conditional cluster
+     * privileges {@link ConditionalClusterPrivilege}
+     *
+     * @param name set of predefined names in {@link DefaultClusterPrivilege} or {@link DefaultConditionalClusterPrivilege} or a valid
+     * cluster action
+     * @return a {@link Tuple} of {@link ClusterPrivilege} and set of predefined fixed conditional cluster privileges
+     * {@link ConditionalClusterPrivilege}
+     */
+    public static Tuple<ClusterPrivilege, Set<ConditionalClusterPrivilege>> resolve(final Set<String> name) {
```

Comment:
There's a few places where this method gets called, but then only `v1()` is extracted from the result.
This feels very much like a problem in the design of this method.
If the caller only wants the privilege object, then why do we resolve everything? and why is it safe for them to throw away the `v2()` ? 

In fact, if I've looked at this correctly, then there's 1 place where we use both parts of the `Tuple` to then construct a `Permission` object, and everywhere else just throws away the ConditionalClusterPrivileges because they just want a `ClusterPrivilege`

Reply:
I share Tim's concern, and I thought how to solve it, but it looks like this is the essence of your proposal...